pkg/proxy/iptables/proxier.go

```diff
@@ -1387,16 +1387,31 @@ func (proxier *Proxier) syncProxyRules() {
 			}
 		}
 
-		// Now write loadbalancing & DNAT rules.
-		n := len(endpointChains)
-		localEndpointChains := make([]utiliptables.Chain, 0)
+		// Firstly, categorize each endpoint into three buckets:
+		//   1. all endpoints that are ready and NOT terminating.
+		//   2. all endpoints that are local, ready and NOT terminating).
+		//   3. all endpoints that are local, serving and terminating.
+		readyEndpointChains := make([]utiliptables.Chain, 0, len(endpointChains))
+		localReadyEndpointChains := make([]utiliptables.Chain, 0, len(endpointChains))
+		localServingTerminatingEndpointChains := make([]utiliptables.Chain, 0, len(endpointChains))
 		for i, endpointChain := range endpointChains {
-			// Write ingress loadbalancing & DNAT rules only for services that request OnlyLocal traffic.
-			if svcInfo.NodeLocalExternal() && endpoints[i].IsLocal {
-				localEndpointChains = append(localEndpointChains, endpointChains[i])
+			if (svc.NodeLocalExternal() || svc.NodeLocalInternal()) && endpoints[i].IsLocal {
```

Comment:
Do we think it's worth doing this for internal or should we scope this to only external? @thockin mentioned previously that we'd want to do this